[dynamo] support dict.clear()

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119197

For code like following:
```python
import torch
def f():
    a = {"a": torch.randn(2, 2)}
    a.clear()
    return a
torch.compile(f, backend="eager", fullgraph=True)()
```

We have a graph break before the pr:
```
torch._dynamo.exc.Unsupported: call_method ConstDictVariable() clear [] {}
```

Test Plan:
Added new tests


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ydwu4 committed:
[dynamo] support dict.clear()

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -241,6 +241,11 @@ def call_method(
             tx.output.side_effects.mutation(self)
             return self.items.pop(Hashable(args[0]))
+        elif name == "clear" and self.mutable_local:
+            assert len(args) == 0 and len(kwargs) == 0
```
```suggestion
            assert len(args) == 0
            assert len(kwargs) == 0
```
makes it easier to know where invariant failed.

ydwu4 committed:
Update on "[dynamo] support dict.clear()"


For code like following:
```python
import torch
def f():
    a = {"a": torch.randn(2, 2)}
    a.clear()
    return a
torch.compile(f, backend="eager", fullgraph=True)()
```

We have a graph break before the pr:
```
torch._dynamo.exc.Unsupported: call_method ConstDictVariable() clear [] {}
```

Test Plan:
Added new tests


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] support dict.clear()"


For code like following:
```python
import torch
def f():
    a = {"a": torch.randn(2, 2)}
    a.clear()
    return a
torch.compile(f, backend="eager", fullgraph=True)()
```

We have a graph break before the pr:
```
torch._dynamo.exc.Unsupported: call_method ConstDictVariable() clear [] {}
```

Test Plan:
Added new tests


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -241,6 +241,12 @@ def call_method(
             tx.output.side_effects.mutation(self)
             return self.items.pop(Hashable(args[0]))
+        elif name == "clear":
+            assert len(args) == 0
+            assert len(kwargs) == 0
+            keys_vt = [key.vt for key in self.items]
+            for key_vt in keys_vt:
+                self.call_method(tx, "pop", [key_vt], {})
```
Why not just:
```
tx.output.side_effects.mutation(self)
return self.items.clear()
```
?

ydwu4 replied:
Yeah, that would work and have better performance. I was hesitating between performance and re-using the code path. I can change to the suggested modification.

ydwu4 committed:
Update on "[dynamo] support dict.clear()"


For code like following:
```python
import torch
def f():
    a = {"a": torch.randn(2, 2)}
    a.clear()
    return a
torch.compile(f, backend="eager", fullgraph=True)()
```

We have a graph break before the pr:
```
torch._dynamo.exc.Unsupported: call_method ConstDictVariable() clear [] {}
```

Test Plan:
Added new tests


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo] support dict.clear()"


For code like following:
```python
import torch
def f():
    a = {"a": torch.randn(2, 2)}
    a.clear()
    return a
torch.compile(f, backend="eager", fullgraph=True)()
```

We have a graph break before the pr:
```
torch._dynamo.exc.Unsupported: call_method ConstDictVariable() clear [] {}
```

Test Plan:
Added new tests


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
